test/TesterInternal/OrleansRuntime/StuckGrainTests.cs

```diff
@@ -0,0 +1,81 @@
+using Orleans.TestingHost;
+using Tester;
+using TestExtensions;
+using UnitTests.GrainInterfaces;
+using Xunit;
+
+namespace UnitTests.StuckGrainTests
+{
+    /// <summary>
+    /// Summary description for PersistenceTest
+    /// </summary>
+    public class StuckGrainTests : OrleansTestingBase, IClassFixture<StuckGrainTests.Fixture>
+    {
+        private class Fixture : BaseTestClusterFixture
+        {
+            protected override TestCluster CreateTestCluster()
+            {
+                GlobalConfiguration.ENFORCE_MINIMUM_REQUIREMENT_FOR_AGE_LIMIT = false;
+                var options = new TestClusterOptions(1);
+                options.ClusterConfiguration.Globals.Application.SetDefaultCollectionAgeLimit(TimeSpan.FromSeconds(3));
+                options.ClusterConfiguration.Globals.CollectionQuantum = TimeSpan.FromSeconds(1);
+
+                return new TestCluster(options);
+            }
+        }
+
+        [Fact, TestCategory("Functional"), TestCategory("ActivationCollection")]
+        public async Task StuckGrainTest_Basic()
+        {
+            var id = Guid.NewGuid();
+            var stuckGrain = GrainClient.GrainFactory.GetGrain<IStuckGrain>(id);
+            var task = stuckGrain.RunForever();
+
+            // Should timeout
+            await Assert.ThrowsAsync<TimeoutException>(() => task.WithTimeout(TimeSpan.FromSeconds(1)));
+
+            var cleaner = GrainClient.GrainFactory.GetGrain<IStuckCleanGrain>(id);
+            await cleaner.Release(id);
+
+            // Should complete now
+            await task.WithTimeout(TimeSpan.FromSeconds(1));
+
+            // wait for activation collection
+            await Task.Delay(TimeSpan.FromSeconds(6)); 
+
+            Assert.False(await cleaner.IsActivated(id), "Grain activation is supposed be garbage collected, but it is still running.");
+        }
+
+        [Fact, TestCategory("Functional"), TestCategory("ActivationCollection")]
+        public async Task StuckGrainTest_StuckDetectionAndForward()
+        {
+            const int nonBlockingCalls = 3;
+
+            var id = Guid.NewGuid();
+            var stuckGrain = GrainClient.GrainFactory.GetGrain<IStuckGrain>(id);
+            var task = stuckGrain.RunForever();
+
+            // Should timeout
+            await Assert.ThrowsAsync<TimeoutException>(() => task.WithTimeout(TimeSpan.FromSeconds(1)));
+
+            for (var i = 0; i < nonBlockingCalls-1; i++)
+            {
+                await Assert.ThrowsAsync<TimeoutException>(
+                    () => stuckGrain.NonBlockingCall().WithTimeout(TimeSpan.FromMilliseconds(500)));
+            }
+
+            // Wait so the first task will reach with DefaultCollectionAge timeout
+            await Task.Delay(TimeSpan.FromSeconds(3));
+
+            // No issue on this one
+            await stuckGrain.NonBlockingCall();
+
+            Assert.Equal(nonBlockingCalls, await stuckGrain.GetNonBlockingCallCounter());
```

Comment:
Good question. The grain will be then still blocked, and I think another thread will be blocked too.

Another solution could be to just removing the activation from the grain directory, but we will not forward pending request (is this really necessary?)